op-batcher/batcher/driver.go

```diff
@@ -313,6 +319,67 @@ func (l *BatchSubmitter) calculateL2BlockRangeToStore(ctx context.Context) (eth.
 // Submitted batch, but it is not valid
 // Missed L2 block somehow.
 
+func (l *BatchSubmitter) checkLeaderElectionBatcherStatus() (bool, error) {
+	// check if L1Block has changed
+	l1tip, err := l.l1Tip(l.shutdownCtx)
+	if err != nil {
+		l.log.Error("Failed to query L1 tip", "error", err)
+		return false, err
+	}
+	if l.lastL1Tip == l1tip {
+		return l.isLeader, nil
+	}
+
+	sysConf, err := bindings.NewSystemConfigCaller(l.Rollup.L1SystemConfigAddress, l.L1Client)
+	if err == nil {
+		batcherHash, err := sysConf.BatcherHash(&bind.CallOpts{BlockNumber: big.NewInt(int64(l1tip.Number)), Context: l.shutdownCtx})
+		if err == nil {
+			if batcherHash[0] == derive.BatchV1Type {
+				l.Config.BatchInboxVersion = derive.BatchV1Type
+			} else if batcherHash[0] == derive.BatchV2Type {
+				l.Config.BatchInboxVersion = derive.BatchV2Type
+			}
+		}
+	}
+
+	if l.Config.BatchInboxVersion == derive.BatchV1Type {
+		return true, nil
+	}
+
+	lebi, err := bindings.NewLeaderElectionBatchInboxCaller(l.Rollup.BatchInboxContractAddr, l.L1Client)
+	if err != nil {
+		l.log.Error("Failed to set up binding to Batch Inbox Contract", "error", err)
+		return false, err
+	}
+	isCurrentLeader, err := lebi.IsCurrentLeader(&bind.CallOpts{Context: l.shutdownCtx}, l.TxManager.From(), big.NewInt(int64(l1tip.Number+1)))
+	if err != nil {
+		l.log.Error("Failed in eth_call to isCurrentLeader", "error", err)
+		return false, err
+	}
+	if !isCurrentLeader {
+		l.isLeader = false
+		return false, nil
+	}
+	l.isLeader = true
+
+	_, leaderSpan, err := lebi.NextBlocksAsLeader(&bind.CallOpts{Context: l.shutdownCtx}, l.TxManager.From(), big.NewInt(int64(l1tip.Number+1)))
+	if err != nil {
+		l.log.Error("Failed in eth_call to nextBlocksAsLeader", "error", err)
+		return false, err
+	}
+	for i, b := range leaderSpan {
+		if !b {
+			if i == 0 {
+				l.state.setLastLeaderBlock(l1tip.Number + 1)
+			} else {
+				l.state.setLastLeaderBlock(l1tip.Number + uint64(i))
```

Comment:
I'm not sure I fully understand. It looks weird to me that for the last leader block ends up the same if `leaderSpan[0]` is the first `false` item and if `leaderSpan[1]` is the first `false` item.

Is the case where  `leaderSpan[0] == false` actually impossible because we would already return on line 361 because `isCurrentLeader` is `false`?

Reply:
It means, at a minimum, that the LeaderElectionBatchInbox contract lies. I can't claim this is _impossible_, because a bad contract could do that. But it is indicative of an incorrect behavior.